### PR TITLE
CI: add appveyor script to build Windows wheels

### PR DESCRIPTION
**Pull request by matthew-brett**
When we release, we will want to build the Windows wheels as well.

See gh-6966

### CODE REVIEW

**Comment by matthew-brett** [comment]
Is there a matplotlib appveyor account?   Can someone give me access so I can set up uploading to the Rackspace wheel container?

**Comment by tacaswell** [comment]
The tests are run on @mdboom 's account.  Is there a way to get org accounts now?

**Comment by matthew-brett** [comment]
I'm not sure - maybe the numpy and astropy accounts I have access to were just personal accounts with suitable names.

**Comment by QuLogic** [comment]
I think you just need to [create a new account for it](http://help.appveyor.com/discussions/questions/1154-appveyor-account-for-github-organizations).

**Comment by matthew-brett** [comment]
@cgohlke - any insight into the error finding the png headers here ?  

https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.2411/job/oo5ajbkacrt9i0ma#L367

**Comment by jenshnielsen** [comment]
I think @janschulz spent a long time debugging libpng issues when setting this up originally.

**Comment by janschulz** [comment]
I had to change a few things in setup*.py to get it building out of the box, but that was more about libfreetype: https://github.com/matplotlib/matplotlib/commits/master/setupext.py?author=janschulz

The png stuff look actually good:

```
copy /Y %LIBRARY_LIB%\libpng_static.lib lib\png.lib
        1 file(s) copied.
del %LIBRARY_LIB%\png.lib
Could Not Find C:\Miniconda35-x64\envs\test-environment\Library\lib\png.lib
```
(the error comes from using conda-forge libpng, which does not ship a unversioned copy of the header file for libpng named `png.lib`. setup should pick up the static `png.lib` file in `lib`)

The visualtests script is only to generate a html file so that the image comparison tests can be viewed in a browser (showing the failures at the top...)

**Comment by janschulz** [comment]
Ok, I think you also need 3aa9eed89c98484142d27d4ae8cc806ea8bdac69 and d53f9027c94ae8b8638a40f47242a8baa46df30b so that the conda dirs are picked up in the script

File: appveyor.yml [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+  # Show the installed packages + versions
+  - conda list
+
+test_script:
+  # Now build the thing..
+  - '%CMD_IN_ENV% python setup.py develop'
+  # these should show no z, png, or freetype dll...
+  - set "DUMPBIN=%VS140COMNTOOLS%\..\..\VC\bin\dumpbin.exe"
+  #- cmd: '"%DUMPBIN%" /DEPENDENTS lib\matplotlib\_png*.pyd'
+  #- cmd: '"%DUMPBIN%" /DEPENDENTS lib\matplotlib\ft2font*.pyd'
```
Comment (janschulz) [inline appveyor.yml]: You might want to uncomment these two to see that they don't include wrong stuff (the next three lines should find prevent that, but to be sure...)

**Comment by janschulz** [comment]
Now there is 
```
======================================================================
FAIL: matplotlib.tests.test_patches.test_wedge_range.test
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Miniconda-x64\envs\test-environment\lib\site-packages\nose\case.py", line 197, in runTest
    self.test(*self.arg)
  File "c:\projects\matplotlib\lib\matplotlib\testing\decorators.py", line 58, in failer
    result = f(*args, **kwargs)
  File "c:\projects\matplotlib\lib\matplotlib\testing\decorators.py", line 261, in do_test
    '(RMS %(rms).3f)'%err)
ImageComparisonFailure: images not close: C:\projects\matplotlib\result_images\test_patches\wedge_range.png vs. C:\projects\matplotlib\result_images\test_patches\wedge_range-expected.png (RMS 0.059)
----------------------------------------------------------------------
Ran 6110 tests in 400.479s
```

which might need https://github.com/matplotlib/matplotlib/commits/b38f558be986e75d31f30cf794409a50f84b906e/lib/matplotlib/tests/test_patches.py?author=janschulz

**Comment by janschulz** [comment]
And maybe some more from https://github.com/matplotlib/matplotlib/pull/5922 ... :-/

**Comment by WeatherGod** [comment]
Current failure:
```
======================================================================
ERROR: matplotlib.tests.test_dviread.test_dviread
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/matplotlib/matplotlib/venv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/matplotlib/matplotlib/venv/lib/python2.7/site-packages/matplotlib-2.0.0b3+110.ga12fcf2-py2.7-linux-x86_64.egg/matplotlib/tests/test_dviread.py", line 62, in test_dviread
    with open(os.path.join(dir, 'test.json')) as f:
IOError: [Errno 2] No such file or directory: u'/home/travis/build/matplotlib/matplotlib/venv/lib/python2.7/site-packages/matplotlib-2.0.0b3+110.ga12fcf2-py2.7-linux-x86_64.egg/matplotlib/tests/baseline_images/dviread/test.json'
```

**Comment by janschulz** [comment]
The windows failure seem to be again some tolerance problems in some tests which were already touched in this PR...

**Comment by matthew-brett** [comment]
OK - after many battles, I think this one is ready now.  All tests passing.

File: setupext.py [hunk 2]
```diff
@@ -1092,15 +1116,52 @@ def do_custom_build(self):
...
+:: REMINDER: fix when changing the version
+copy %FREETYPE%\\objs\\{vc20xx}\\{xXX}\\freetype261.lib %FREETYPE%\\objs\\.libs\\libfreetype.lib 
+if errorlevel 1 (
+  rem This is a py27 version, which has a different location for the lib file :-/
+  copy %FREETYPE%\\objs\\win32\\{vc20xx}\\freetype261.lib %FREETYPE%\\objs\\.libs\\libfreetype.lib
+)
+"""
+            from setup_external_compile import fixproj, prepare_build_cmd, VS2010, X64, tar_extract
+            # Note: freetype has no build profile for 2014, so we don't bother...
+            vc = 'vc2010' if VS2010 else 'vc2008'
```
Comment (cgohlke) [inline setupext.py]: This is broken on Python 3.5, which uses VS2015. Better raise an Exception.

Comment (matthew-brett) [inline setupext.py]: Build seems to be working : https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.2436/job/akufhhatuycx8tps - is it not running this block?

**Comment by cgohlke** [comment]
Compared to the matplotlib 1.x wheels, these wheels are missing the gtkagg extensions for Python 2.7 and the Visual C++ 2015 runtime DLLs for Python 3.5.

File: setupext.py (hunk 2 above)
Comment (cgohlke) [inline setupext.py]: Looks like you got lucky with linking a static library built with VS2010 into a DLL otherwise built with VS2015.

**Comment by matthew-brett** [comment]
Christoph - would you consider sharing your recipes to get something automated working on appveyor?

**Comment by cgohlke** [comment]
To build freetype with VS2015 it should be possible to convert the 2010 solution and build it with msbuild. Try:
```
devenv freetype-2.6.1\builds\windows\vc2010\freetype.sln /upgrade
msbuild.exe freetype-2.6.1\builds\windows\vc2010\freetype.sln /t:Clean;Build /p:Configuration="Release";Platform=x64
```
Do not initialize the Windows 7 SDK, only VS2015.

To bundle the VS2015 runtime DLLs I run
```CMD
echo [gui_support] > setup.cfg
echo windowing = True >> setup.cfg
echo [rc_options] >> setup.cfg
echo backend = TkAgg >> setup.cfg
echo [package_data] >> setup.cfg
echo dlls = True >> setup.cfg
::<snip>
copy /Y /B "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\redist\x64\Microsoft.VC140.CRT\MSVCP140.DLL" lib\matplotlib\
copy /Y /B "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\redist\x64\Microsoft.VC140.CRT\CONCRT140.DLL" lib\matplotlib\
::<snip>
python.exe setup.py bdist_wheel
del /S /Q lib\matplotlib\*.dll
```

For the GTK2 agg backend I am using the [GTK+ 2.22 distributions](http://ftp.gnome.org/pub/GNOME/binaries/) and my own builds of [PyGTK](http://www.lfd.uci.edu/~gohlke/pythonlibs/#pygtk). Nothing automated, just setting PATH, INCLUDE, LIB, and PKG_CONFIG_PATH environment variables.
Not sure if it is worth providing the GTK2 agg backend for matplotlib 2.x. I have not tried so far.

**Comment by janschulz** [comment]
All comments from @cgohlke probably also apply to the master version of the appveyor script :-/

The MS runtime libs are probably not included because I (and the appveyor scripts) use conda and conda has them included when python is installed, so no error was raised... :-/

**Comment by cgohlke** [comment]
Re MS runtime: Appveyor will probably never fail because the DLLs are installed as part of Visual Studio 2015. Many user systems are missing the DLLs, which is a nuisance, so the DLLs are included in the matplotlib wheels.

**Comment by NelleV** [comment]
Hi @matthew-brett 
Do you mind making a check list of what is left to be done on this PR? I think that would help us move forward.

**Comment by dstansby** [comment]
Moving to `2.1` since this isn't really a bug fix

**Pull request closed without merging**